### PR TITLE
fix(cheatcodes): Correct `expectRevert` behavior

### LINKED ISSUE #3437: Test function early returns after `vm.expectRevert`
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (d7733ee 2022-10-03T00:06:06.841223Z)

### What command(s) is the bug in?

forge test

### Operating System

macOS (Apple Silicon)

### Describe the bug

```solidity
pragma solidity 0.8.4;

import "forge-std/Test.sol";

contract ContractTest is Test {
    function setUp() public {}

    function rever() internal {
        revert();
    }

    function testExample() public {
        vm.expectRevert();
        rever();

        rever();
        console.log("Does not print");
    }
}
```
Running `forge test -vvv` passes this test and also doesn't print "Does not print" on console. In summary, after `vm.expectRevert()` the test just executes the call and does not run the remaining test.

### LINKED ISSUE #3723: Forge doesn't work with multiple "vm.expectRevert" when testing free functions
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (cb925b1 2022-11-21T00:09:19.970948Z)

### What command(s) is the bug in?

forge test

### Operating System

macOS (Apple Silicon)

### Describe the bug

Given the following test:

```solidity
function testMultipleReverts() public {
    vm.expectRevert();
    revert();

    vm.expectRevert();
    console2.log("Do not revert");
}
```

Forge will consider the test to be passed, even if it should be failed, because there is no second revert.

### LINKED ISSUE #4832: bug(forge): `vm.expectRevert` without external call makes `assertEq` pass
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

_No response_

### What command(s) is the bug in?

forge test

### Operating System

macOS (Apple Silicon)

### Describe the bug

This test passes.

```solidity
    function testFun() public {
        assertEq(uint256(1), 2);

        vm.expectRevert();
        revert();
    }
```

```
Running 1 test for test/Contract.t.sol:ContractTest
[PASS] testFun() (gas: 15068)
Logs:
  Error: a == b not satisfied [uint]
        Left: 1
       Right: 2

Test result: ok. 1 passed; 0 failed; finished in 283.88µs
```

I know `vm.expectRevert` is not intended to be used on anything but external calls, but I still like using it for testing free functions.

### PR DESCRIPTION
**Pull request by Evalir**
** This is a big breaking change**

## Motivation

Closes #3723 , Closes #3437 , Closes #4832 .

`expectRevert` currently has incorrect behavior: it matches reverts at the _test_ level, not at the _next call_ level. This means that reverting cheatcodes were being incorrectly catched by `expectRevert` (we're supposed to bypass checks for cheatcode calls) and also let users use the cheatcode incorrectly as illustrated in #3723.

## Solution and Behavior

`expectRevert` now only works on the next `CALL`. Cheatcode calls are now ignored properly, _even if those cheatcode calls revert_. To illustrate, these are examples of now legal/illegal behavior:

```solidity
// OK
function testExpectRevertString() public {
    Reverter reverter = new Reverter();
    cheats.expectRevert("revert");
    reverter.revertWithMessage("revert");
}

// FAIL
function testFailRevertNotOnImmediateNextCall() public {
    Reverter reverter = new Reverter();
    // expectRevert should only work for the next call. However,
    // we do not inmediately revert, so,
    // we fail.
    cheats.expectRevert("revert");
    reverter.doNotRevert();
    reverter.revertWithMessage("revert");
}

// FAIL
function testFailCheatcodeRevert() public {
  // This expectRevert is hanging, as the next cheatcode call is ignored.
  vm.expectRevert();
  vm.fsMetadata(fakePath) // try to go to some non-existent path to cause a revert
}
```

### On tests marked as "not working properly"

As you can see, the diff is extremely minimal in terms of the change itself. This has uncovered a few tests which are not working as expected anymore—mainly because they were relying on the broken `expectRevert` behavior. Some of them seem to be bugs—mainly the `fs`/`parsing` related ones.

NOTE: This PR is done—just needs a coordinated forge-std release once this goes on main, which won't be soon.

### CODE REVIEW

**Comment by Evalir** [comment]
Working on the forge-std tests—the behavior change also breaks some tests there

File: evm/src/executor/inspector/cheatcodes/mod.rs [hunk 1]
```diff
@@ -821,6 +821,15 @@ where
 
         // If the depth is 0, then this is the root call terminating
         if data.journaled_state.depth() == 0 {
+            // See if there's a dangling expectRevert that should've been matched.
+            if self.expected_revert.is_some() {
+                return (
+                    InstructionResult::Revert,
+                    remaining_gas,
+                    "Expected revert was left dangling".encode().into(),
```
Comment (mds1) [inline evm/src/executor/inspector/cheatcodes/mod.rs:829]: Nit on phrasing to make it more clear, open to alternatives
```suggestion
                    "A `vm.expectRevert` was left dangling".encode().into(),
```

Comment (Evalir) [reply evm/src/executor/inspector/cheatcodes/mod.rs:829]: Maybe we could hint at how it behaves? Something like:
```suggestion
                    "A `vm.expectRevert`was left dangling. Make sure that calls you expect to revert are external".encode().into(),
```

wdyt?

Comment (mds1) [reply evm/src/executor/inspector/cheatcodes/mod.rs:829]: that sound great to me 👌

Comment (Evalir) [reply evm/src/executor/inspector/cheatcodes/mod.rs]: done in https://github.com/foundry-rs/foundry/pull/4945/commits/3bc04ef3776f93468bf52026cb703fe6992d3285 !

**Pull request merged**